### PR TITLE
BinarySearchTree [Python]

### LINKED ISSUE #547: Binary Search Tree [Python]
#### This issue is about [Algo/DS Name](link to resource for the Algo/DS).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by avsingh999**
Fixes #547

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by singhpratyush** [comment]
Hi! Thanks for showing interest in contributing to this repository.

Please make sure that you have ticked the points mentioned in PR description correctly.

Thanks again!

File: binary_search_tree/BinarySearchTree.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+class Node(object):
+# Node for Binary search Tree and set the Node
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
@@ -1,5 +1,5 @@
 class Node(object):
-# Node for Binary search Tree and set the Node
+    # Node for Binary search Tree and set the Node
     def __init__(self, data):
         self.data = data
         self.left = None
```

File: binary_search_tree/BinarySearchTree.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    # find the minimum
+    def minright(node):
+        if node.left is None:
+            return node
+        else:
+            node = Bst.min_right(node.left)
+        return node
+
+
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
@@ -46,9 +46,8 @@
             node = Bst.min_right(node.left)
         return node
 
-
     # delete a node
-    def delete(root,data):
+    def delete(root, data):
         if root is None:
             return root
         if data < root.data:
```

File: binary_search_tree/BinarySearchTree.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    # print inorder
+    def inorder(root):
+        if root is not None:
+            Bst.inorder(root.left)
+            print(root.data)
+            Bst.inorder(root.right)
+
+
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
@@ -76,7 +76,6 @@
             print(root.data)
             Bst.inorder(root.right)
 
-
     # print preorder
     def preorder(root):
         if root is not None:
```

File: binary_search_tree/BinarySearchTree.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    # print preorder
+    def preorder(root):
+        if root is not None:
+            print(root.data)
+            Bst.preorder(root.left)
+            Bst.preorder(root.right)
+
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
@@ -84,7 +84,6 @@
             Bst.preorder(root.left)
             Bst.preorder(root.right)
 
-
     # print postorder
     def postorder(root):
         if root is not None:
```

File: binary_search_tree/BinarySearchTree.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    print("****** InOrder ******")
+    Bst.inorder(root)
+    print("****** PreOrder ******")
+    Bst.preorder(root)
+    print("****** PostOrder ******")
+    Bst.postorder(root)
+    print("***** search Node ******")
+    temo = Bst.search(root, 28)
+    if temo is not None:
+        print("node search==> ", temo.data )
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmp0xwmz3s9/binary_search_tree/BinarySearchTree.py
@@ -113,7 +113,7 @@
     print("***** search Node ******")
     temo = Bst.search(root, 28)
     if temo is not None:
-        print("node search==> ", temo.data )
+        print("node search==> ", temo.data)
     else:
         print("node is not present")
     print("***** delete Node ******")
```

File: binary_search_tree/BinarySearchTree.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        self.data = data
+        self.left = None
+        self.right = None
+
+
+class Bst(object):
+    # insert data in tree
+    def insert(node, data):
+        if node is None:
+            # create a new node
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpfqi8ue1b/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmpfqi8ue1b/binary_search_tree/BinarySearchTree.py
@@ -47,7 +47,7 @@
         return node
 
     # delete a node
-    def delete(root,data):
+    def delete(root, data):
         if root is None:
             return root
         if data < root.data:
```

File: binary_search_tree/BinarySearchTree.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            node = Bst.search(node.right, data)
+        # data found
+        elif data == node.data:
+            return node
+        return node
+
+    # find the minimum
+    def minright(node):
+        if node.left is None:
+            return node
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpfqi8ue1b/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmpfqi8ue1b/binary_search_tree/BinarySearchTree.py
@@ -89,6 +89,8 @@
             print(root.data)
 
 # check the method
+
+
 def main():
     root = None
     root = Bst.insert(root, 5)
```

File: binary_search_tree/BinarySearchTree.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    if node is None:
+        # create a new node
+        node = Node(data)
+        return node
+    else:
+        # go to the left child
+        if data <= node.data:
+            node.left = insert(node.left, data)
+        # got to the right child
+        else:
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
@@ -6,6 +6,8 @@
         self.right = None
 
 # insert data in tree
+
+
 def insert(node, data):
     if node is None:
         # create a new node
```

File: binary_search_tree/BinarySearchTree.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+# find the minimum
+def minright(node):
+    if node.left is None:
+        return node
+    else:
+        node = minright(node.left)
+    return node
+
+# delete a node
+def delete(root, data):
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
@@ -21,6 +21,8 @@
     return node
 
 # seraching data
+
+
 def search(node, data):
     # if data not present
     if node is None:
```

File: binary_search_tree/BinarySearchTree.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        # if left or right child is Not None find the the minimum node in right child
+        temp = minright(root.right)
+        root.data = temp.data
+        # recursive delete the minimum node in right child
+        root.right = delete(root.right, temp.data)
+    return root
+
+# print inorder
+def inorder(root):
+    if root is not None:
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
@@ -37,6 +37,8 @@
         return node
 
 # find the minimum
+
+
 def minright(node):
     if node.left is None:
         return node
```

File: binary_search_tree/BinarySearchTree.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+# check the method
+
+
+def main():
+    root = None
+    root = insert(root, 5)
+    root = insert(root, 11)
+    root = insert(root, 3)
+    root = insert(root, 1)
```
Comment (singhpratyush) [inline binary_search_tree/BinarySearchTree.py]: The code does not comply to PEP8.

*Origin: PEP8Bear, Section: `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
+++ b/tmp/tmppakk8uf0/binary_search_tree/BinarySearchTree.py
@@ -45,6 +45,8 @@
     return node
 
 # delete a node
+
+
 def delete(root, data):
     if root is None:
         return root
```

**Comment by singhpratyush** [comment]
@avsingh999: Please mention the issue number and make sure that you are assigned to it. Thanks.

**Comment by aashutoshrathi** [comment]
@avsingh999 Please add its entry to README

**Comment by aashutoshrathi** [comment]
Please rebase. It's good to go @singhpratyush. Take a look

**Comment by singhpratyush** [comment]
Hey! This pull request is being rebased automatically. Please DO NOT push while rebase is in progress or your changes would be lost!

**Comment by singhpratyush** [comment]
Automated rebase was successful!

File: binary_search_tree/BinarySearchTree.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    root = insert(root, 30)
+    root = insert(root, 35)
+    print("****** InOrder ******")
+    inorder(root)
+    print("****** PreOrder ******")
+    preorder(root)
+    print("****** PostOrder ******")
+    postorder(root)
+    print("***** search Node ******")
+    temo = search(root, 28)
```
Comment (aashutoshrathi) [inline binary_search_tree/BinarySearchTree.py]: Better if you use `temp`

**Pull request merged**